datafusion/expr-common/src/signature.rs

```diff
@@ -125,6 +125,11 @@ pub enum TypeSignature {
     /// Fixed number of arguments of numeric types.
     /// See <https://docs.rs/arrow/latest/arrow/datatypes/enum.DataType.html#method.is_numeric> to know which type is considered numeric
     Numeric(usize),
+    /// Fixed number of arguments of string types.
```

Comment:
It's ok for now, but it feels pretty inflexible.
Why have special handling for functions that take only strings or only numbers?
What about functions that accept one (any) string and one (any) number? (eg `substr`)

The Numeric and String could be "a type class"

```rust
// strpos
Signature::from_type_classes(vec![
  TypeClass:String,
  TypeClass:String,
]);

// hypothetical substr
Signature::from_type_classes(vec![
  TypeClass:String,
  TypeClass:Numeric,
  TypeClass:Numeric,
]);
```

This would be more flexible. However, aren't we just defining the "logical types" (https://github.com/apache/datafusion/issues/11513) this way?
Maybe we could introduce such a concept into the codebase and simply start using it incrementally?

cc @alamb  @notfilippo

Reply:
Yes, I agree that this could be the start of effectively defining a logical `String` type

Coercion actually seems like a pretty good place to start doing so, when I think about it

Reply:
I know, that is why I didn't introduce this signature before, but works on #12622 first. This PR is mainly fixing the regression.